packages/drivers/driver-base/src/driverUtils.ts

```diff
@@ -103,3 +105,39 @@ export async function promiseRaceWithWinner<T>(
 		});
 	});
 }
+
+export function validateMessages(
+	reason: string,
+	messages: ISequencedDocumentMessage[],
+	from: number,
+	logger: ITelemetryLoggerExt,
+) {
+	if (messages.length !== 0) {
+		const start = messages[0].sequenceNumber;
+		const length = messages.length;
+		const last = messages[length - 1].sequenceNumber;
+		if (start !== from) {
+			logger.sendErrorEvent({
+				eventName: "OpsFetchViolation",
+				reason,
+				from,
+				start,
+				last,
+				length,
+			});
+			messages.length = 0;
+		}
+		if (last + 1 !== from + length) {
+			logger.sendErrorEvent({
+				eventName: "OpsFetchViolation",
+				reason,
+				from,
+				start,
+				last,
+				length,
+			});
+			// we can do better here by finding consecutive sub-block and return it
+			messages.length = 0;
```

Comment:
I know it's moved code from ODSP and ODSP does not hit this problems. My question is about - can we do better for AFR which struggles for last 6+ months to fix an issue with their ops having gaps. While it's not a good long term strategy (we need fixes on service side), if client can limp along with partial results, this might be good change to make. That said, we can do it in separate PR and it's not very clear if it will help (as retrying whole range should have same results, even if less efficient).